src/Riok.Mapperly/Descriptors/MappingBuilders/StringToEnumMappingBuilder.cs

```diff
@@ -43,51 +46,105 @@ private static EnumFromStringSwitchMapping BuildEnumFromStringSwitchMapping(
     )
     {
-        var ignoredTargetMembers = ctx.Configuration.Enum.IgnoredTargetMembers.ToHashSet(SymbolTypeEqualityComparer.FieldDefault);
-        var explicitMappings = BuildExplicitValueMappings(ctx);
-
+        var enumMemberMappings = BuildEnumMemberMappings(ctx);
         // from string => use an optimized method of Enum.Parse which would use slow reflection
         // however we currently don't support all features of Enum.Parse yet (ex. flags)
         // therefore we use Enum.Parse as fallback.
         var fallbackMapping = BuildFallbackParseMapping(ctx, genericEnumParseMethodSupported);
-        var members = ctx.SymbolAccessor.GetAllFields(ctx.Target);
-        if (fallbackMapping.FallbackMember != null)
+        if (fallbackMapping.FallbackExpression is MemberAccessExpressionSyntax fallbackMember)
         {
             // no need to explicitly map fallback value
-            members = members.Where(x => fallbackMapping.FallbackMember.ConstantValue?.Equals(x.ConstantValue) != true);
+            enumMemberMappings = enumMemberMappings.Where(IsNotEquivalentTo(fallbackMember));
         }
 
-        EnumMappingDiagnosticReporter.AddUnmatchedTargetIgnoredMembers(ctx, ignoredTargetMembers);
         return new EnumFromStringSwitchMapping(
             ctx.Source,
             ctx.Target,
-            members,
             ctx.Configuration.Enum.IgnoreCase,
-            fallbackMapping,
-            explicitMappings
+            enumMemberMappings,
+            fallbackMapping
         );
     }
 
+    private static IEnumerable<EnumMemberMapping> BuildEnumMemberMappings(MappingBuilderContext ctx)
+    {
+        var namingStrategy = ctx.Configuration.Enum.NamingStrategy;
+
+        var ignoredTargetMembers = ctx.Configuration.Enum.IgnoredTargetMembers.ToHashSet(SymbolTypeEqualityComparer.FieldDefault);
+        EnumMappingDiagnosticReporter.AddUnmatchedTargetIgnoredMembers(ctx, ignoredTargetMembers);
+
+        var fields = ctx.SymbolAccessor.GetFieldsExcept(ctx.Target, ignoredTargetMembers);
+        var explicitValueMappings = BuildExplicitValueMappings(ctx);
+        var customNameMappings = ctx.BuildCustomNameStrategyMappings(ctx.Target);
+
+        foreach (var field in fields)
+        {
+            // source.Value1
+            var targetSyntax = MemberAccess(FullyQualifiedIdentifier(ctx.Target), field.Name);
+
+            if (explicitValueMappings.TryGetValue(field, out var explicitMappings))
+            {
+                foreach (var explicitMapping in explicitMappings)
+                {
+                    // "explicit_value1" => source.Value1
+                    yield return new EnumMemberMapping(explicitMapping, targetSyntax);
+                }
+                continue;
+            }
+
+            ExpressionSyntax sourceSyntax;
+            if (namingStrategy is not EnumNamingStrategy.MemberName && customNameMappings.TryGetValue(field, out var customNameMapping))
```

Comment:
Nitpick: i think the readability could be improved with early yield returns + continue statements.

Reply:
Good point, updated with early yield returns and continue statements for better readability.